File: main/java/com/websarva/wings/android/dowasurememo/UpdateMemo.java
package com.websarva.wings.android.dowasurememo;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;

import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.Toast;

public class UpdateMemo extends AppCompatActivity {

    private Databasehelper _helper;
    private String _category = "update1";

    private int indexCounter=1;
    int tagId=1;
    String table="update1";
    Context context=UpdateMemo.this;

    LayoutInflater inflater;
    LinearLayout linearLayout;
    LinearLayout llUpdateLayout;
    LinearLayout llUpdateInputform;
    LinearLayout llUpdateTitle;
    LinearLayout llUpdateDeadline;

    EditText etUpdateTitle;
    EditText etUpdateYear;
    EditText etUpdateMonth;
    EditText etUpdateDay;
    ImageButton btDelete;
    ImageButton btDateSelect;

    String strUpdateTitle;
    String strUpdateYear;
    String strUpdateMonth;
    String strUpdateDay;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_update_memo);

        Intent intent = getIntent();
        llUpdateLayout = findViewById(R.id.ll_update_layout);

        _helper=new Databasehelper(getApplicationContext());
        SQLiteDatabase db=_helper.getWritableDatabase();
        String sqlSelect="SELECT * FROM update1";
        Cursor cursor=db.rawQuery(sqlSelect,null);
        cursor.moveToFirst();
        while (cursor.moveToNext()) {
            int i = cursor.getColumnIndex("_id");
            tagId = cursor.getInt(i);
            inflater = LayoutInflater.from(getApplicationContext());
            llUpdateLayout = findViewById(R.id.ll_update_layout);
            llUpdateInputform=(LinearLayout)inflater.inflate(R.layout.update_inputform,null);
            llUpdateLayout.addView(llUpdateInputform);

            llUpdateTitle = llUpdateInputform.findViewById(R.id.ll_update_title);
            llUpdateDeadline = llUpdateTitle.findViewById(R.id.ll_update_deadline);

            etUpdateTitle = llUpdateTitle.findViewById(R.id.et_update_title);
            etUpdateYear = llUpdateDeadline.findViewById(R.id.et_update_year);
            etUpdateMonth = llUpdateDeadline.findViewById(R.id.et_update_month);
            etUpdateDay = llUpdateDeadline.findViewById(R.id.et_update_day);
            btDelete = llUpdateDeadline.findViewById(R.id.bt_delete);
            btDelete.setOnClickListener
                    (new DeleteButton(UpdateMemo.this,llUpdateLayout,llUpdateInputform,table));
            btDateSelect = llUpdateDeadline.findViewById(R.id.bt_date_select);
            btDateSelect.setOnClickListener(new UpdateMemo.DatePicker());


            etUpdateTitle.setTag(tagId);
            etUpdateYear.setTag(tagId);
            etUpdateMonth.setTag(tagId);
            etUpdateDay.setTag(tagId);
            btDelete.setTag(tagId);



            i = cursor.getColumnIndex("updatetitle");
            strUpdateTitle = cursor.getString(i);

            i = cursor.getColumnIndex("updateyear");
            strUpdateYear = cursor.getString(i);

            i = cursor.getColumnIndex("updatemonth");
            strUpdateMonth = cursor.getString(i);

            i = cursor.getColumnIndex("updateday");
            strUpdateDay = cursor.getString(i);


            try {
                etUpdateTitle.setText(strUpdateTitle);
                EditEventListener etListener=new EditEventListener(etUpdateTitle,UpdateMemo.this);
                etUpdateTitle.addTextChangedListener(etListener);
            } catch (NullPointerException e) {
                strUpdateTitle = "";
            }

            try {
                etUpdateYear.setText(strUpdateYear);
                EditEventListener etListener2=new EditEventListener(etUpdateYear,UpdateMemo.this);
                etUpdateYear.addTextChangedListener(etListener2);
            } catch (NullPointerException e) {
                strUpdateYear = "";
            }

            try {
                etUpdateMonth.setText(strUpdateMonth);
                EditEventListener etListener3=new EditEventListener(etUpdateMonth,UpdateMemo.this);
                etUpdateMonth.addTextChangedListener(etListener3);
            } catch (NullPointerException e) {
                strUpdateMonth = "";
            }

            try {
                etUpdateDay.setText(strUpdateDay);
                EditEventListener etListener3=new EditEventListener(etUpdateDay,UpdateMemo.this);
                etUpdateDay.addTextChangedListener(etListener3);
            } catch (NullPointerException e) {
                strUpdateDay = "";
            }
        }

        DatabaseControl control=new DatabaseControl(context,table);
        indexCounter=control.GetIndexCounter();
        Log.d("main",""+indexCounter);


    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater=getMenuInflater();
        inflater.inflate(R.menu.optionmenu,menu);
        return super.onCreateOptionsMenu(menu);
    }

    @RequiresApi(api = Build.VERSION_CODES.Q)
    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.option_add:
                inflater = LayoutInflater.from(getApplicationContext());
                llUpdateInputform=(LinearLayout)inflater.inflate(R.layout.update_inputform,null);
                llUpdateLayout.addView(llUpdateInputform);

                llUpdateTitle=llUpdateInputform.findViewById(R.id.ll_update_title);
                llUpdateDeadline=llUpdateTitle.findViewById(R.id.ll_update_deadline);

                etUpdateTitle=llUpdateTitle.findViewById(R.id.et_update_title);
                etUpdateYear=llUpdateDeadline.findViewById(R.id.et_update_year);
                etUpdateMonth=llUpdateDeadline.findViewById(R.id.et_update_month);
                etUpdateDay=llUpdateDeadline.findViewById(R.id.et_update_day);
                btDelete=llUpdateDeadline.findViewById(R.id.bt_delete);
                btDelete.setOnClickListener
                        (new DeleteButton(UpdateMemo.this,llUpdateLayout,llUpdateInputform,table));
                btDateSelect = llUpdateDeadline.findViewById(R.id.bt_date_select);
                btDateSelect.setOnClickListener(new UpdateMemo.DatePicker());

                etUpdateTitle.setTag(indexCounter);
                etUpdateYear.setTag(indexCounter);
                etUpdateMonth.setTag(indexCounter);
                etUpdateDay.setTag(indexCounter);
                btDelete.setTag(indexCounter);

                EditEventListener etListener=new EditEventListener(etUpdateTitle,UpdateMemo.this);
                etUpdateTitle.addTextChangedListener(etListener);
                EditEventListener etListener2=new EditEventListener(etUpdateYear,UpdateMemo.this);
                etUpdateYear.addTextChangedListener(etListener2);
                EditEventListener etListener3=new EditEventListener(etUpdateMonth,UpdateMemo.this);
                etUpdateMonth.addTextChangedListener(etListener3);
                EditEventListener etListener4=new EditEventListener(etUpdateDay,UpdateMemo.this);
                etUpdateDay.addTextChangedListener(etListener4);

                tagId=indexCounter;
                String str="";

                DatabaseControl control=new DatabaseControl(context,table);
                control.DatabaseDelete(tagId);

                String column1="updatetitle";
                String column2="updateyear";
                String column3="updatemonth";
                String column4="updateday";

                DatabaseControl control2=new DatabaseControl
                        (context,table,tagId,_category,str);
                control2.DatabaseInsert(column1,column2,column3,column4);

                indexCounter++;
                control.IndexCounterUpdate(indexCounter);



        }
        return super.onOptionsItemSelected(item);
    }

    public class DatePicker implements View.OnClickListener {

        @Override
        public void onClick(View v) {
            LinearLayout parentView= (LinearLayout) v.getParent();
            EditText etYear=parentView.findViewById(R.id.et_update_year);
            EditText etMonth=parentView.findViewById(R.id.et_update_month);
            EditText etDay=parentView.findViewById(R.id.et_update_day);


            Toast.makeText
                    (UpdateMemo.this,
                            "西暦の変更は左上に薄く表示されている西暦の箇所をタップしてください",
                            Toast.LENGTH_LONG).show();

            DatePickerFragment datePicker =
                    new DatePickerFragment(etYear,etMonth,etDay);
            datePicker.show(getSupportFragmentManager(), "datePicker");

        }
    }

}

File: main/java/com/websarva/wings/android/dowasurememo/UpdateMemo.kt
package com.websarva.wings.android.dowasurememo

import android.content.Context
import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.widget.EditText
import android.widget.ImageButton
import android.widget.LinearLayout
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.FragmentManager

/**
 * 更新期限を入力するUpdateメモのクラス
 *
 * @author nakayama
 * @version 1.0
 */
class UpdateMemo : AppCompatActivity() {
    private val context: Context = this@UpdateMemo
    private var inflater: LayoutInflater? = null
    private var llUpdateLayout: LinearLayout? = null
    private var llUpdateInputform: LinearLayout? = null
    private var llUpdateTitle: LinearLayout? = null
    private var llUpdateDeadline: LinearLayout? = null
    private var etUpdateTitle: EditText? = null
    private var etUpdateYear: EditText? = null
    private var etUpdateMonth: EditText? = null
    private var etUpdateDay: EditText? = null
    private var btDelete: ImageButton? = null
    private var btDateSelect: ImageButton? = null
    private var strUpdateTitle: String? = null
    private var strUpdateYear: String? = null
    private var strUpdateMonth: String? = null
    private var strUpdateDay: String? = null
    var manager: FragmentManager? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_update_memo)
        inflater = LayoutInflater.from(applicationContext)
        llUpdateLayout = findViewById(R.id.ll_update_layout)
        llUpdateInputform = inflater.inflate(R.layout.update_inputform, null) as LinearLayout
        /**
         * データベースの列名の配列
         */
        val columnNames = listOf("updatetitle", "updateyear", "updatemonth", "updateday")
        /**
         * カレンダーによる日付選択用のFragmentManager型の変数
         */
        manager = supportFragmentManager

        //データベースからデータを取り出して、レイアウトを作成する処理
        val control = DatabaseControl(context, TABLE, columnNames, manager)
        control.selectDatabase(llUpdateLayout)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.optionmenu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    @RequiresApi(api = Build.VERSION_CODES.Q)
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.option_add -> {
                inflater = LayoutInflater.from(applicationContext)
                llUpdateInputform = inflater.inflate(R.layout.update_inputform, null) as LinearLayout
                llUpdateLayout!!.addView(llUpdateInputform)
                llUpdateTitle = llUpdateInputform!!.findViewById(R.id.ll_update_title)
                llUpdateDeadline = llUpdateTitle.findViewById(R.id.ll_update_deadline)
                btDelete = llUpdateDeadline.findViewById(R.id.bt_delete)
                btDelete.setOnClickListener(DeleteButton(this@UpdateMemo, llUpdateLayout, llUpdateInputform, TABLE))
                btDateSelect = llUpdateDeadline.findViewById(R.id.bt_date_select)
                btDateSelect.setOnClickListener(DatePickerListener(context, manager, TABLE))
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onPause() {
        super.onPause()

        //データベースにある全てのデータを削除
        val control = DatabaseControl(context, TABLE)
        control.deleteAllDatabase()

        //メモの文字列を取得してデータベースにインサートする
        for (i in 0 until llUpdateLayout!!.childCount) {
            val linearLayout = llUpdateLayout!!.getChildAt(i) as LinearLayout
            etUpdateTitle = linearLayout.findViewById(R.id.et_update_title)
            etUpdateYear = linearLayout.findViewById(R.id.et_update_year)
            etUpdateMonth = linearLayout.findViewById(R.id.et_update_month)
            etUpdateDay = linearLayout.findViewById(R.id.et_update_day)
            strUpdateTitle = etUpdateTitle.getText().toString()
            strUpdateYear = etUpdateYear.getText().toString()
            strUpdateMonth = etUpdateMonth.getText().toString()
            strUpdateDay = etUpdateDay.getText().toString()
            val control2 = DatabaseControl(context, TABLE, i, _CATEGORY, strUpdateTitle, strUpdateYear, strUpdateMonth, strUpdateDay)
            control2.insertDatabaseFourColumns("updatetitle", "updateyear", "updatemonth", "updateday")
        }
        finish()
    }

    companion object {
        private const val _CATEGORY = "UPDATE1"

        /**
         * データベースのテーブル名
         */
        private const val TABLE = "update1"
    }
}